src/measure_theory/pi.lean

```diff
@@ -0,0 +1,231 @@
+by { induction l with i l ih, exact dirac punit.star, exact (μ i).prod ih }
+
+@[simp] lemma tprod_nil (μ : Π i, measure (π i)) : measure.tprod [] μ = dirac punit.star := rfl
+
+@[simp] lemma tprod_cons (i : δ) (l : list δ) (μ : Π i, measure (π i)) :
+  measure.tprod (i :: l) μ = (μ i).prod (measure.tprod l μ) := rfl
+
+instance sigma_finite_tprod (l : list δ) (μ : Π i, measure (π i)) [∀ i, sigma_finite (μ i)] :
+  sigma_finite (measure.tprod l μ) :=
+begin
+  induction l with i l ih,
+  { rw [tprod_nil], apply_instance },
+  { rw [tprod_cons], resetI, apply_instance }
+end
+
+lemma tprod_tprod (l : list δ) (μ : Π i, measure (π i)) [∀ i, sigma_finite (μ i)]
+  {s : Π i, set (π i)} (hs : ∀ i, is_measurable (s i)) :
+  measure.tprod l μ (set.tprod l s) = (l.map (λ i, (μ i) (s i))).prod :=
+begin
+  induction l with i l ih, { simp },
+  simp_rw [tprod_cons, set.tprod, prod_prod (hs i) (is_measurable.tprod l hs), map_cons,
+    prod_cons, ih]
+end
+
+lemma tprod_tprod_le (l : list δ) (μ : Π i, measure (π i)) [∀ i, sigma_finite (μ i)]
+  (s : Π i, set (π i)) : measure.tprod l μ (set.tprod l s) ≤ (l.map (λ i, (μ i) (s i))).prod :=
+begin
+  induction l with i l ih, { simp [le_refl] },
+  simp_rw [tprod_cons, set.tprod, map_cons, prod_cons],
+  refine (prod_prod_le _ _).trans _, exact ennreal.mul_left_mono ih
+end
+
+end tprod
+
+section encodable
+
+open list measurable_equiv
+variables [encodable ι]
+
+/-- The product measure on an encodable finite type, defined by mapping `measure.tprod` along the
+  equivalence `measurable_equiv.pi_measurable_equiv_tprod`.
+  The definition `measure_theory.measure.pi` should be used instead of this one. -/
+def pi' : measure (Π i, α i) :=
+measure.map (tprod.elim' encodable.mem_sorted_univ) (measure.tprod (encodable.sorted_univ ι) μ)
+
+lemma pi'_pi [∀ i, sigma_finite (μ i)] {s : Π i, set (α i)}
+  (hs : ∀ i, is_measurable (s i)) : measure.pi' μ (pi univ s) = ∏ i, μ i (s i) :=
+begin
+  have hl := λ i : ι, encodable.mem_sorted_univ i,
+  have hnd := @encodable.sorted_univ_nodup ι _ _,
+  rw [pi', map_apply (measurable_tprod_elim' hl) (is_measurable.pi_fintype (λ i _, hs i)),
+    elim_preimage_pi hnd, tprod_tprod _ μ hs, ← list.prod_to_finset _ hnd],
+  congr' with i, simp [hl]
+end
+
+lemma pi'_pi_le [∀ i, sigma_finite (μ i)] {s : Π i, set (α i)} :
+  pi' μ (pi univ s) ≤ ∏ i, μ i (s i) :=
+begin
+  have hl := λ i : ι, encodable.mem_sorted_univ i,
+  have hnd := @encodable.sorted_univ_nodup ι _ _,
+  apply ((pi_measurable_equiv_tprod hnd hl).symm.map_apply (pi univ s)).trans_le,
+  dsimp only [pi_measurable_equiv_tprod, tprod.pi_equiv_tprod, coe_symm_mk, equiv.coe_fn_symm_mk],
+  rw [elim_preimage_pi hnd],
+  refine (tprod_tprod_le _ _ _).trans_eq _,
+  rw [← list.prod_to_finset _ hnd],
+  congr' with i, simp [hl]
+end
+
+end encodable
+
+lemma pi_caratheodory :
+  measurable_space.pi ≤ (outer_measure.pi (λ i, (μ i).to_outer_measure)).caratheodory :=
+begin
+  refine supr_le _, intros i s hs,
```

Comment:
can you do one tactic per line, here and below, unless you're closing a goal? (I mean, the current lines 203 or 204 or 220 are fine but 197 or 198 or 217 or 225 or 226 are not)

Reply:
Will do. Could you make a PR to the [style guide](https://leanprover-community.github.io/contribute/style.html) with the one tactic per line rule?

Reply:
Just so I understand: at the time of your comment, line 226 closed a goal, so was that fine?

Reply:
It wasn't fine because it is not a one-liner to close a goal, it comes in a sequence of tactics that should all be on its own line. I'll make a PR to the style guide to try to clarify this (and maybe we'll get feedback that will converge in another direction, but in any case it will be good to have a guide to a more uniform style).